be/src/storage/rowset_column_update_state.cpp

```diff
@@ -106,79 +100,78 @@ Status RowsetColumnUpdateState::_load_upserts(Rowset* rowset, uint32_t idx) {
     }
 
+    std::shared_ptr<Chunk> chunk_shared_ptr = ChunkHelper::new_chunk(pkey_schema, DEFAULT_CONTAINER_SIZE);
     // iterators for each update segments;
-    std::vector<ChunkIteratorPtr> itrs;
-
-    std::shared_ptr<Chunk> chunk_shared_ptr;
-    // if we want to preload update cache, create chunk with update columns
-    if (_enable_preload_column_mode_update_data) {
-        _update_chunk_cache[idx] = ChunkHelper::new_chunk(update_schema, DEFAULT_CONTAINER_SIZE);
-        chunk_shared_ptr = ChunkHelper::new_chunk(update_schema_with_keys, DEFAULT_CONTAINER_SIZE);
-        ASSIGN_OR_RETURN(itrs, rowset->get_update_file_iterators(update_schema_with_keys, &stats));
-    } else {
-        _update_chunk_cache[idx] = ChunkHelper::new_chunk(update_schema, 0);
-        chunk_shared_ptr = ChunkHelper::new_chunk(pkey_schema, DEFAULT_CONTAINER_SIZE);
-        ASSIGN_OR_RETURN(itrs, rowset->get_update_file_iterators(pkey_schema, &stats));
-    }
+    ASSIGN_OR_RETURN(std::vector<ChunkIteratorPtr> itrs, rowset->get_update_file_iterators(pkey_schema, &stats));
 
     if (itrs.size() != rowset->num_update_files()) {
         std::string err_msg = fmt::format("itrs.size {} != num_update_files {}, tablet_id: {}", itrs.size(),
                                           rowset->num_update_files(), _tablet_id);
         DCHECK(false) << err_msg;
         return Status::InternalError(err_msg);
     }
-    auto chunk = chunk_shared_ptr.get();
-    auto& dest = _upserts[idx];
-    auto& dest_chunk = _update_chunk_cache[idx];
-    auto col = pk_column->clone();
-    auto itr = itrs[idx].get();
-    if (itr != nullptr) {
-        col->reserve(DEFAULT_CONTAINER_SIZE);
-        while (true) {
-            chunk->reset();
-            auto st = itr->get_next(chunk);
-            if (st.is_end_of_file()) {
-                break;
-            } else if (!st.ok()) {
-                return st;
-            } else {
-                PrimaryKeyEncoder::encode(pkey_schema, *chunk, 0, chunk->num_rows(), col.get());
-                if (_enable_preload_column_mode_update_data) {
-                    for (int i = pk_columns.size(); i < update_columns_with_keys.size(); i++) {
-                        _memory_usage += chunk->columns()[i]->byte_size();
-                        dest_chunk->columns()[i - pk_columns.size()]->append(*chunk->columns()[i]);
-                    }
+
+    // alloc first BatchPKsPtr
+    auto header_ptr = std::make_shared<BatchPKs>();
+    header_ptr->upserts = pk_column->clone();
+    header_ptr->start_idx = start_idx;
+    for (uint32_t idx = start_idx; idx < rowset->num_update_files(); idx++) {
+        _update_chunk_cache[idx] = ChunkHelper::new_chunk(update_schema, 0);
+        header_ptr->offsets.push_back(header_ptr->upserts->size());
+        auto chunk = chunk_shared_ptr.get();
+        auto col = pk_column->clone();
+        auto itr = itrs[idx].get();
+        if (itr != nullptr) {
+            col->reserve(DEFAULT_CONTAINER_SIZE);
```

Comment:
We already defined a `DEFAULT_CHUNK_SIZE` in execution. You can merge the two const variable.

Reply:
Done